Sources/BSON/CodableNumber.swift

```diff
@@ -0,0 +1,193 @@
+            if let exact = Self(exactly: v) {
+                self = exact
+                return
+            }
+        case let .double(v):
+            if let exact = Self(exactly: v) {
+                self = exact
+                return
+            }
+        default:
+            break
+        }
+        return nil
+    }
+
+    /// By default, just try casting the number to a `BSONValue`. Types
+    /// where that will not work provide their own implementation of the
+    /// `bsonValue` computed property.
+    internal var bsonValue: BSONValue? {
+        self as? BSONValue
+    }
+}
+
+extension Int: CodableNumber {
+    internal var bsonValue: BSONValue? {
+        BSON(self).bsonValue
+    }
+}
+
+extension Int32: CodableNumber {}
+extension Int64: CodableNumber {}
+
+extension Int8: CodableNumber {
+    internal var bsonValue: BSONValue? {
+        // Int8 always fits in an Int32
+        Int32(exactly: self)
+    }
+}
+
+extension Int16: CodableNumber {
+    internal var bsonValue: BSONValue? {
+        // Int16 always fits in an Int32
+        Int32(exactly: self)
+    }
+}
+
+extension UInt8: CodableNumber {
+    internal var bsonValue: BSONValue? {
+        // UInt8 always fits in an Int32
+        Int32(exactly: self)
+    }
+}
+
+extension UInt16: CodableNumber {
+    internal var bsonValue: BSONValue? {
+        // UInt16 always fits in an Int32
+        Int32(exactly: self)
+    }
+}
+
+extension UInt32: CodableNumber {
+    internal var bsonValue: BSONValue? {
+        // try an Int32 first
+        if let int32 = Int32(exactly: self) {
+            return int32
+        }
+        // otherwise, will always fit in an Int64
+        return Int64(exactly: self)
+    }
+}
+
+extension UInt64: CodableNumber {
+    internal var bsonValue: BSONValue? {
+        if let int32 = Int32(exactly: self) {
+            return int32
+        }
+        if let int64 = Int64(exactly: self) {
+            return int64
+        }
+        if let double = Double(exactly: self) {
+            return double
+        }
+        // we could consider trying a Decimal128 here. However,
+        // it's not clear how we could support decoding something
+        // stored as Decimal128 back to a UInt64 without access
+        // to libbson internals.
+        return nil
+    }
+}
+
+extension UInt: CodableNumber {
+    internal var bsonValue: BSONValue? {
+        if let int32 = Int32(exactly: self) {
+            return int32
+        }
+        if let int64 = Int64(exactly: self) {
+            return int64
+        }
+        if let double = Double(exactly: self) {
+            return double
+        }
+        // we could consider trying a Decimal128 here. However,
+        // it's not clear how we could support decoding something
+        // stored as Decimal128 back to a UInt without access
+        // to libbson internals.
+        return nil
+    }
+}
+
+/// Override the default initializer due to a runtime assertion that fails
```

Comment:
I wonder if we still still need to do this? same for Float below. I think those comments are really old, probably pre Swift 5.

Reply:
done. seems to work fine.